WooCommerce/Classes/ViewRelated/Notifications/DefaultReviewsDataSource.swift

```diff
@@ -0,0 +1,202 @@
+import Foundation
+import UIKit
+import Yosemite
+
+
+/// Default implementation of the ReviewsDataSource, dequeeues and
+/// populates cells to render the Product Review list
+///
+final class DefaultReviewsDataSource: NSObject, ReviewsDataSource {
+
+    // MARK: - Private properties
+
+    /// Product Reviews
+    ///
+    private lazy var reviewsResultsController: ResultsController<StorageProductReview> = {
+        let storageManager = ServiceLocator.storageManager
+        let descriptor = NSSortDescriptor(keyPath: \StorageProductReview.dateCreated, ascending: false)
+
+        return ResultsController<StorageProductReview>(storageManager: storageManager,
+                                                       sectionNameKeyPath: "normalizedAgeAsString",
+                                                       matching: self.filterPredicate,
+                                                       sortedBy: [descriptor])
+    }()
+
+    /// Products
+    ///
+    private lazy var productsResultsController: ResultsController<StorageProduct> = {
+        let storageManager = ServiceLocator.storageManager
+        let descriptor = NSSortDescriptor(keyPath: \StorageProduct.productID, ascending: true)
+
+        return ResultsController<StorageProduct>(storageManager: storageManager,
+                                                       matching: sitePredicate,
+                                                       sortedBy: [descriptor])
+    }()
+
+    /// Predicate to filter only Product Reviews that are either approved or on hold
+    ///
+    private lazy var filterPredicate: NSPredicate = {
+        let statusPredicate = NSPredicate(format: "statusKey ==[c] %@ OR statusKey ==[c] %@",
+                                          ProductReviewStatus.approved.rawValue,
+                                          ProductReviewStatus.hold.rawValue)
+
+        return  NSCompoundPredicate(andPredicateWithSubpredicates: [sitePredicate, statusPredicate])
+    }()
+
+    /// Predicate to entities that belong to the current store
+    ///
+    private lazy var sitePredicate: NSPredicate = {
+        return NSPredicate(format: "siteID == %lld",
+                          ServiceLocator.stores.sessionManager.defaultStoreID ?? Int.min)
```

Comment:
True. But I kind of see it as yet another global dependency, similar to the StoresManager. At least, it is a global dependency at runtime. To be honest, I do not see the benefit of injecting it (yes, it would make some tests slightly easier to write, but in the end, I think that eventually we are going to need to inject more than just the site id into the sessionManager that we use for testing) 

Also, look at the call site when this class is constructed. Right now, it is (in ReviewsViewController):
```
private let viewModel = ReviewsViewModel(data: DefaultReviewsDataSource())
```

If we inject the siteID into both the viewModel and this datasource, that line would be:

```
private let viewModel = ReviewsViewModel(siteID: siteID, data: DefaultReviewsDataSource(siteID: siteID))
```

I am leaning towards leaving it as it is. What do you think?

Reply:
Oh, and there is one more thing (which might actually be the deal breaker). These classes have to support changing the store, if we inject the siteID via the constructor we wouldn't be able to support that, as far as I understand

Reply:
ah I see, the support for changing stores seems like a deal-breaker. if this view controller is per site, I think another benefit of DI here is also to ensure the site ID remains the same for its lifetime